### PR TITLE
feat(products): add product service implementation

### PR DESCRIPTION
**Pull request by Davy-Dray**

### CODE REVIEW

File: src/products/controller/product.controller.ts [hunk 1]
```diff
@@ -0,0 +1,92 @@
+import { Body, Controller, Delete, Get, Param, Patch, Post, UseGuards } from "@nestjs/common";
+import { ProductService } from "../service/product.service";
+import { JwtAuthGuard } from "~/auth/guards/auth.guard";
+import { RolesGuard } from "~/auth/guards/role.guard";
+import { CreateProductDTO } from "../dto/create-product.dto";
```
Comment (iamveekthorr) [inline src/products/controller/product.controller.ts:5]: Try to arrage your imports to follow a particular order

File: src/products/dto/create-product.dto.ts [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+    @IsNotEmpty()
+    description: string;
+
+    @IsNumber()
+    @IsNotEmpty()
+    price: number;
+
+    @IsNumber()
+    @IsOptional()
+    stock?: number;
```
Comment (iamveekthorr) [inline src/products/dto/create-product.dto.ts:26]: we could reword this to say 'quantity' so it's more descriptive

File: src/products/service/product.service.ts [hunk 3]
```diff
@@ -0,0 +1,176 @@
...
+        }).lean();
+
+        if (!product) {
+            throw new AppError(
+                `Product not found`,
+                HttpStatus.NOT_FOUND,
+            );
+        }
+
+        const store = product.store as unknown as { owner: Types.ObjectId };
```
Comment (iamveekthorr) [inline src/products/service/product.service.ts:140]: Please visit this

File: src/products/service/product.service.ts [hunk 4]
```diff
@@ -0,0 +1,176 @@
...
+
+        const isStoreOwner = store.owner.toString() === userId;
+
+        if (!isStoreOwner) {
+            throw new AppError(
+                `Forbidden`,
+                HttpStatus.FORBIDDEN,
+            );
+        }
```
Comment (iamveekthorr) [inline src/products/service/product.service.ts:149]: Why didn't you just pass the userId into the query so when you don't find the product, you can as well just throw an error and move on

**Review by iamveekthorr — COMMENTED** [review]
Overall i think you should look at the way the files are imported and order them starting with npm modules and then imoorted modules and then the files that belong to that module. Asides that the PR looks good to go with a few updates

**Review by iamveekthorr — CHANGES_REQUESTED** [review]
Please check the initial comments made

File: src/products/service/product.service.ts (hunk 4 above)
Comment (Davy-Dray) [reply src/products/service/product.service.ts:149]: that way , for some reasons I don't know , it returns "null" for the product

Comment (Davy-Dray) [reply src/products/service/product.service.ts:149]: const product = await this.productModel.findOne({
            _id: productId,
            
            store: {
              _id: userId,
            },
            
          });

File: src/products/service/product.service.ts [hunk 5]
```diff
@@ -0,0 +1,175 @@
...
+                HttpStatus.NOT_FOUND,
+            );
+        }
+
+        if (store.owner.toString() !== userId) {
+            throw new AppError(
+                `Forbidden`,
+                HttpStatus.FORBIDDEN,
+            );
+        }
```
Comment (iamveekthorr) [inline src/products/service/product.service.ts:111]: Good catch

**Review by iamveekthorr — CHANGES_REQUESTED** [review]
Please check comments

**Review by iamveekthorr — APPROVED** [review]
